### ISSUE TITLE
new constraint

### ISSUE THREAD

**Issue by rgleason** [rgleason/weather_routing_pi#84]
https://routing.luckgrib.com/intro/constraints.html

Take to Starbord or port point constraint.
See luckgrib

**Comment by rgleason** [comment]
@sebastien-rosset   Sebastien you've got this planned should I close it?

**Comment by sebastien-rosset** [comment]
Yes I have started to work on it. I suggest closing this issue when I am done you have tested the capability.


 There is some work in opencpn core, the plug-in API and then in the WR plugin. I think some discussion is needed with @bdcat who suggested that maybe this should be only a WR plugin feature.
I'm not really sure if this will make it into the 12 release.

On a related note, I've started implementing a smarter routing algorithm that could potentially take into consideration multiple constraints, including but not limited to land, depth, buoys, beacons, port/starboard side, etc.
One idea I have in mind that when a user is actively drawing a route, open CPN core would render the shortest path that satisfies all the constraints. Currently, when you draw a route, open CPN renders the rhumb line or the GC line 
 Open CPN will happily go over land, shallow depth, rocks, wrecks and other hazards.

With this new feature, open CPN would calculate a path that avoids the obstacles, as the user moves the mouse when drawing the route. So for example in an extreme case if the starting point of the route is somewhere near Panama in the Pacific Ocean and you move the mouse to the Bahamas, it could potentially show you a route segment that takes you around Cape Horn. 
It's fairly complex but it's also a really exciting feature to work on. 
With all these use cases in mind and more, worth having these constraints in opencpn core. I.e., it's not just about the WR plug-in, but it's also about providing a routing capability directly in open CPN.

**Comment by sebastien-rosset** [comment]
I also want to point out that I need to do a lot of experimentation, and in particular figure out whether it's possible to do just in time routing calculation or if it's too expensive. If it turns out to be expensive, then the shortest path calculation will be done on demand as opposed to dynamically when the user is actively drawing a route.


I'm planning to start with an R tree algorithm which is in log N and supports high resolution. Then depending on how it goes, I may also implement a distant field which provides O(1) constant time, but comes with additional complexity because it's a grid with a trade-off between CPU usage and memory usage.

**Comment by rgleason** [comment]
> There is some work in opencpn core, the plug-in API and then in the WR plugin. I think some discussion is needed with @bdbcat who suggested that maybe this should be only a WR plugin feature.
> I'm not really sure if this will make it into the 12 release.

@sebastien-rosset  Is there any testing  (@quinton-hoole) that we  could do that would help with acceptance of your important PR into OpenCPN?  What branch of your opencpn fork should we use?

**Comment by rgleason** [comment]
> I think some discussion is needed with @bdbcat who suggested that maybe this should be only a WR plugin feature.

Here is a good place to have that discussion, I think.
[Weather_Routing and Opencpn Development Planning (api, versions, etc.) #4485  ](https://github.com/OpenCPN/OpenCPN/discussions/4485)

**Comment by DavidKroc** [comment]
If possible, it would be nice to consider difference of a (race) buoy you have to "left" (port or starboard) with a (race) buoy you have to "round" (port or starboard). With a left buoy, it doesn't matter how far the boat is from the buoy. While a round buoy is mandatory point of crossings: you have to pass near it.

It would also be nice the start could be a line between 2 buoys instead of a single point, so that the routing chooses the best position behind this line to start from instead of a fixed point. And same thing for the arrival.

**Comment by rgleason** [comment]
1. Starboard rounding, any distance
2. Starboard rounding, within X distance
3. Port rounding, any distance
4. Port rounding, within X distance
5. Gates/Start pair of Port/Starboard bouys.
